ocaml/xapi/xapi_pool_patch.ml

```diff
@@ -284,36 +285,47 @@ let pool_patch_upload_handler (req: Request.t) s _ =
 
         Http_svr.headers s (Http.http_200_ok ());
 
-        (match req.Request.content_length with
-         | None -> ()
-         | Some size -> assert_space_available size);
-
-        read_in_and_check_patch req.Request.content_length s new_path;
-	
-        try
-          let r = create_patch_record ~__context ~path:new_path (get_patch_info new_path) in
-          Db.Task.set_result ~__context ~self:task_id ~value:(Ref.string_of r)   
-        with Db_exn.Uniqueness_constraint_violation (_, _, uuid) ->
-          (* patch already uploaded.  if the patch file has been cleaned, then put this one in its place. 
-             otherwise, error *)
-          debug "duplicate patch with uuid %s found." uuid;
-          let patch_ref = Db.Pool_patch.get_by_uuid ~__context ~uuid in
-          let old_path = Db.Pool_patch.get_filename ~__context ~self:patch_ref in
-          debug "checking for file %s.  If it doesn't exist new patch will replace it." old_path;
-          if Sys.file_exists old_path
-          then
-            begin
-              Unixext.unlink_safe new_path;
-              raise (Api_errors.Server_error(Api_errors.patch_already_exists, [uuid])) 
-            end
-          else
-            begin
-              let stat = Unix.stat new_path in
-              let size = Int64.of_int stat.Unix.st_size in
-                Db.Pool_patch.set_filename ~__context ~self:patch_ref ~value:new_path;
-                Db.Pool_patch.set_size ~__context ~self:patch_ref ~value:size;
-                Db.Task.set_result ~__context ~self:task_id ~value:(Ref.string_of patch_ref)   
-            end
+        let size = match req.Request.content_length with
+        | Some size -> size
+        | None ->
+          warn "no content-length HTTP header, I'm assuming this patch is 1GiB in size.";
+          Int64.(mul 1024L (mul 1024L 1024L)) in
+
+        (* Slaves without VDI support will write the patch to Pool_patch.filename
+           in their local filesystem, so it needs to be a sensible path. *)
+        let legacy_path = patch_dir ^ "/" ^ (Uuid.to_string (Uuid.make_uuid ())) in
+
+        (* Ensure we have 1x the space for the upload, 1x the space for the
+           de-signatured patch, 1x for unpacking and 1x for luck *)
+        let required_space = Int64.mul 4L size in
+        let sR = Helpers.choose_patch_sr ~__context ~host:(Helpers.get_localhost ~__context) in
+        Sm_fs_ops.with_new_fs_vdi __context ~name_label:"Unknown patch"
+          ~name_description:"" ~sR ~required_space ~_type:`user ~sm_config:[]
+        (fun vdi patch_dir ->
+          let new_path = patch_dir ^ "/patch" in
+          read_in_and_check_patch req.Request.content_length s new_path;
```

Comment:
If I'm reading this aright, this is still going onto the dom0 filesystem rather than the VDI.
Could we use the VDI here?
Also, it uses a hardcoded path; this means each patch will overwrite the previous one (so we don't use more space for each new patch) but there could be trouble if there are multiple simultaneous pool_patch_upload calls in progress.